### ISSUE TITLE
[Win64] -Duselongdouble perls built using mingw-w64 port of gcc-10.2.0 prone to crashing.

### ISSUE THREAD

**Issue by sisyphus** [Perl/perl5#18363]
My OS is Windows 7.
The compilers I've used are:
https://github.com/brechtsanders/winlibs_mingw/releases/download/10.2.0-11.0.0-9.0.0-r3/winlibs-i686-posix-dwarf-gcc-10.2.0-llvm-11.0.0-mingw-w64-8.0.0-r3.7z
and
https://github.com/brechtsanders/winlibs_mingw/releases/download/10.2.0-11.0.0-9.0.0-r3/winlibs-x86_64-posix-seh-gcc-10.2.0-llvm-11.0.0-mingw-w64-8.0.0-r3.7z

No problems building perl (both nvtype eq 'double' and nvtype eq 'long double') with the former of those two compilers (which is 32-bit).
But using the latter (64-bit) compiler presents crashes under certain conditions -  when and only when perl is built with nvtype of "long double".
When these crashes occur, they present a pop-up box stating that 'perl.exe has stopped working', with no additional diagnostics.

There are no issues if I switch to using mingw-w64 ports of gcc-8.3.0 that were built by the Strawberry Perl developers.

**Description**

The bug prevents the 'make test' step of the perl build process from occurring. As soon as the first of the test files are about to be run, the process crashes.
Also, having installed perl (via 'make install'), attempts at installing additional modules via the 'cpan' utility will crash.
In fact, merely trying to load CPAN.pm causes a crash.
Instead, one has to install modules manually:
perl Makefile.PL
make install

Attempts to run 'make test' again result in the crash.
Also, Inline::C is unusable, with a crash occurring during the attempted compilation of the C code.

**Steps to Reproduce**

This 6-line script drawn from code in CPAN.pm  (search that file for "SAVEOUT") prints out <code>OK so far</code> and then delivers the crash.

<code>{</code>
<code>my $x = *SAVEOUT; # avoid warning</code>
<code>open($x, ">&STDOUT");</code>
<code>print "OK so far\n";</code>
<code>}</code>
<code>print "Still OK\n";</code>

This 5-line variant is fine:

<code>{</code>
<code>open(my $x, ">&STDOUT");</code>
<code>print "OK so far\n";</code>
<code>}</code>
<code>print "Still OK\n";</code>

Having output:

<code>OK so far</code>
<code>Still OK</code>

it then terminates normally.
That also demonstrates how I avoid having CPAN.pm crash upon loading when I try to use the cpan utility to install a module.
However, the same crash then subsequently occurs elsewhere - at a point that I haven't yet located.
I assume that the "subsequent" crash occurs for the same or similar reason ... but that's just assumption.

**Perl configuration**

```
Summary of my perl5 (revision 5 version 33 subversion 4) configuration:
   
  Platform:
    osname=MSWin32
    osvers=6.1.7601
    archname=MSWin32-x64-multi-thread-ld
    uname=''
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=define
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags =' -DWIN32 -DWIN64 -fdiagnostics-color=never -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO -D__USE_MINGW_ANSI_STDIO -fno-strict-aliasing -mms-bitfields'
    optimize='-s -O2'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='10.2.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long long'
    ivsize=8
    nvtype='long double'
    nvsize=16
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags ='-s -L"C:\perl-5.33.4-1020\lib\MSWin32-x64-multi-thread-ld\CORE" -L"C:\_64\winlibs-gcc-1020\mingw64\lib"'
    libpth=C:\_64\winlibs-gcc-1020\mingw64\lib C:\_64\winlibs-gcc-1020\mingw64\lib C:\_64\winlibs-gcc-1020\mingw64\x86_64-w64-mingw32\lib C:\_64\msys_1020\1.0\local\lib C:\_64\winlibs-gcc-1020\mingw64\lib\gcc\x86_64-w64-mingw32\10.2.0
    libs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs= -lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=
    so=dll
    useshrplib=true
    libperl=libperl533.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-mdll -s -L"C:\perl-5.33.4-1020\lib\MSWin32-x64-multi-thread-ld\CORE" -L"C:\_64\winlibs-gcc-1020\mingw64\lib"'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_IMPLICIT_CONTEXT
    PERL_IMPLICIT_SYS
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_LONG_DOUBLE
    USE_PERLIO
    USE_PERL_ATOF
  Built under MSWin32
  Compiled at Nov 21 2020 11:05:42
  @INC:
    C:/perl-5.33.4-1020/site/lib/MSWin32-x64-multi-thread-ld
    C:/perl-5.33.4-1020/site/lib
    C:/perl-5.33.4-1020/lib/MSWin32-x64-multi-thread-ld
    C:/perl-5.33.4-1020/lib
```

Cheers,
Rob

**Comment by sisyphus** [comment]
I still don't know much about this bug. (I don't know if it's a perl bug, or a gcc-10.2.0 bug, or a bug in the mingw-w64 64-bit porting of gcc-10.2.0, or a bug in the way this particular 64-bit compiler was built.)
I just wanted to add here that there are no issues when a USE_QUADMATH build is done using this gcc-10.2.0 compiler.
The problem exists only with the 64-bit compiler and only then if the nvtype is long double.

Cheers,
Rob

**Comment by sisyphus** [comment]
I've found a solution - it's a pretty simple fix, but I don't know whether it's actually *"the"* solution.

In perl.h, in current blead, we have:
```
#if defined(USE_QUADMATH) && defined(__MINGW64__)
   /* 64-bit build, mingw-w64 compiler only */
   typedef NVTYPE NV __attribute__ ((aligned(8)));
#else
...
```
My solution is to change the condition to:
```
#if (defined(USE_LONG_DOUBLE) || defined(USE_QUADMATH)) && defined(__MINGW64__)
```
That does the job nicely.
However, that's a condition that applies to all 64-bit mingw-w64 compilers building a USE_LONG_DOUBLE build of perl - though it's probably only my 64-bit mingw-w64 port of gcc-10.2.0 (built by Brecht Sanders) that actually requires this condition.
Here's a list of other 64-bit mingw-w64 ports of gcc I have:
gcc-4.9.2
gcc-5.3.0
gcc-6.3.0
gcc-7.3.0
gcc-8.1.0
gcc-8.3.0

Two of those were built by StrawberryPerl Project, the remainder built by mingw-w64 project.
I believe that none of them require this fix - at least, they were fine without it when I last used them to build a USE_LONG_DOUBLE perl .
I have checked that with each of those compilers, building current blead (with USE_LONG_DOUBLE &&) with my solution in place does not cause any breakage that is caught by the test suite.

Is this solution acceptable ?
If so, I'll provide a PR that applies it.

It's certainly an acceptable solution from my POV, and it's a solution I'll be using locally until something better comes along.

But it does occur to me that maybe I'm just applying bandaids to something that should be fixed differently inside the perl source.
I'm also a little uneasy about such a "cover all" fix being applied to situations where it's not needed, and without any indication of whether it was needed or not.
I'll listen, with interest, to any guidance provided here.

In case it helps, I'm providing the configuration info for both the gcc-10.2.0 port and the gcc-8.1.0 port.
(This is the configuration info as provide by <code>gcc -v</code> output.

[gcc_config_1020.txt](https://github.com/Perl/perl5/files/5805829/gcc_config_1020.txt)
[gcc_config_810.txt](https://github.com/Perl/perl5/files/5805832/gcc_config_810.txt)

Cheers,
Rob

**Comment by sisyphus** [comment]
I didn't pursue this in the lead up to the release of perl-5.34.0, but I'd like to pick this up again as it's a continuing issue for me.

To recap - the problem is that 'long double' builds of perl are unusable when built with 64-bit mingw-w64 ports of gcc-9 and later.
I have now checked with 9.3.0 (runtime 7.0), 10.2.0 (runtime 8.0), 10.3.0 (runtime 8.0) and 11.1.0 (runtime 9.0) - all as built by Brecht Sanders and available from https://winlibs.com, and it's the same problem.

Currently in perl.h we have:
```
#if defined(USE_QUADMATH) && defined(__MINGW64__)
   /* 64-bit build, mingw-w64 compiler only */
   typedef NVTYPE NV __attribute__ ((aligned(8)));
#else
...
```
The solution that I've been using is to replace that condition:
```
#if defined(USE_QUADMATH) && defined(__MINGW64__)
```
with the condition:
```
#if (defined(USE_LONG_DOUBLE) || defined(USE_QUADMATH)) && defined(__MINGW64__)
```
It's working fine for me, and it's the only solution I have.
Is that acceptable ?
Or do I need to add <code>&& __MINGW64_VERSION_MAJOR >= 9</code> to those conditions, given that gcc-8.x.x and earlier are, apparently immune from the issue.
UPDATE: On checking, the correct condition would be (as intimated above) <code>&& __MINGW64_VERSION_MAJOR >= 7</code>

I would think that some explanatory comments would also be required, but I'm loath to expend further energy on this until I know exactly what's required.

Cheers,
Rob

**Comment by sisyphus** [comment]
Fixed by the merging of #19437
